### PR TITLE
W6 Pesto Assignments

### PR DESCRIPTION
**Pull request by siddhantdalvi3**

### CODE REVIEW

File: Week-6/W6Exercise6.1/Index.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+const maxSumContiguousSubArray = (_arr) => {
+    let maxSoFar = Number.MIN_VALUE, maxEndingHere = 0, start = 0, end = 0, temp = 0;
+    for (let i = 0; i < _arr.length; i++) {
```
Comment (kapilggg10) [inline Week-6/W6Exercise6.1/Index.js:3]: Explain this loop in laymen terms.

File: Week-6/W6Exercise6.3/Index.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+const sortArr = (arr) => {
+    let low = 0, high = arr.length - 1, mid = 0, temp = 0;
+    while (mid <= high) {
+        if (arr[mid] === 0) {
+            temp = arr[mid];
```
Comment (kapilggg10) [inline Week-6/W6Exercise6.3/Index.js:5]: can you think of swapping two elements in js in a better way?

File: Week-6/W6Exercise6.3/Index.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+const sortArr = (arr) => {
+    let low = 0, high = arr.length - 1, mid = 0, temp = 0;
+    while (mid <= high) {
```
Comment (kapilggg10) [inline Week-6/W6Exercise6.3/Index.js:3]: Explanation for this loop?

File: Week-6/W6Exercise6.3/Index.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+            mid++;
+        }
+        else {
+            temp = arr[mid];
+            arr[mid] = arr[high];
+            arr[high] = temp;
+            high--;
+        }
+    }
+    return arr;
```
Comment (kapilggg10) [inline Week-6/W6Exercise6.3/Index.js:21]: any other approaches you can think??

File: Week-6/W6Exercise6.5/Index.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+const findPairWithDifference = (arr, difference) => {
+    const map = new Map();
```
Comment (kapilggg10) [inline Week-6/W6Exercise6.5/Index.js:2]: Its O(N) in terms of memory and time complexity, can you think of better approach?

**Review by kapilggg10 — APPROVED** [review]
Think of some counterpart solutions which can be costlier or cheaper.

File: Week-6/W6Exercise6.1/Index.js (hunk 1 above)
Comment (siddhantdalvi3) [reply Week-6/W6Exercise6.1/Index.js:3]: In loop

- Adding the elements
- Checking if the added value is greater than maxSoFar which is min value present in number, assuming if the array could also consist of only negative values.
- If it is greater then, maxSofar is assigned value of total added till now.
- start & end, variables are used for returning the start and end of the sub array
- checking if total added so far is still less than zero, reassigning it to zero, and temp value to i+1

File: Week-6/W6Exercise6.3/Index.js (hunk 3 above)
Comment (siddhantdalvi3) [reply Week-6/W6Exercise6.3/Index.js:3]: Sorting 0, 1, 2

- 4 variables are initialized, mid for 1, low for 0 and high for 2, and temp for swapping values
- 3 conditions checks to check if the current element is 0, 1 or 2
- If element is 0, then swapping element to the starting side of the array, where low variables of array comes in play which starts from 0, if the element is 0 then element is swapped, low is incremented coz now its sorted and mid is incremented to move onto next value.
- if element is 1, there is no need for swapping, mid is incremented and loop moves on to next value.
- If element is 2, element is swapped with ending side of the loop, and high is deremented

File: Week-6/W6Exercise6.3/Index.js (hunk 2 above)
Comment (siddhantdalvi3) [reply Week-6/W6Exercise6.3/Index.js:5]: `[arr[mid], arr[low]] = [arr[low], arr[mid]]`

Would use this from now on

File: Week-6/W6Exercise6.3/Index.js (hunk 4 above)
Comment (siddhantdalvi3) [reply Week-6/W6Exercise6.3/Index.js:21]:
```
function customSort(arr: number[]): number[] {
  let result: number[] = [];
  let zero: number[] = [];
  let one: number[] = [];
  let two: number[] = [];
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === 0) {
      zero.push(arr[i]);
    } else if (arr[i] === 1) {
      one.push(arr[i]);
    } else {
      two.push(arr[i]);
    }
  }
  result = result.concat(zero, one, two);
  return result;
}

const sortArr = (arr: number[]) => {
  for (let i = 0; i < arr.length; i++) {
    for (let j = i + 1; j < arr.length; j++) {
      if (arr[i] > arr[j]) {
        [arr[i], arr[j]] = [arr[j], arr[i]];
      }
    }
  }
  return arr;
};
```

File: Week-6/W6Exercise6.5/Index.js (hunk 5 above)
Comment (siddhantdalvi3) [reply Week-6/W6Exercise6.5/Index.js:2]:
```
const findPairWithDifference = (arr, difference) => {
  const _arr = [];
  while (arr.length > 0) {
    const first = arr.pop();
    const second = first + difference;
    if (_arr.includes(second)) {
      return 1;
    }
    _arr.push(first);
  }
  return 0;
};
```

**Review by kapilggg10 — APPROVED** [review]

**Pull request merged**